keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+- The only requirement for deployments is that the components can connect to
+  the API server. Direct communication is not needed, but in some cases
+  can be optionally used to improve performance.
+
+Using a single object is an intentional simplification compared to the
+PersistentVolume/PersistentVolumeClaim model for volumes where the additional
+PV object was used to capture status. That model allowed operations like
+pre-provisioning volumes and then having Kubernetes bind those to claims that
+get created later. For resources, the resource driver can and must handle such
+pre-provisioning internally. Kubernetes wouldn't know how to match
+pre-provisioned resources against claims because it has no understanding about the
+parameters.
+
+The entire state of a resource can be determined by looking at its
+ResourceClaim (see [API below](#api) for details), for example:
+
+- It is allocated if and only if `ResourceClaimStatus.Allocated` is non-nil and
+  points to the `AllocationResult`.
+
+- It is in use if and only if `ResourceClaimStatus.ReservedFor` contains one or
+  more users.  It does not matter whether those users, usually pods, are
+  currently running because that could change at any time.
+
+- A resource is no longer needed when it has a `DeletionTimestamp`. It must not
+  be freed yet when it is still in use.
+
+Some of the race conditions that need to be handled are:
+
+- A ResourceClaim gets created and deleted again while the resource driver
+  starts allocating it. Before it actually starts doing anything, the
+  driver adds a finalizer. Either adding the finalizer or removing the
+  ResourceClaim win. If the driver wins, it continues with allocation
+  and can either complete or abort the operation when it notices the non-nil
+  DeletionTimestamp. Otherwise, allocation gets aborted immediately.
+
+  What this avoids is the situation where an allocation succeed without having
+  an object where the result can be stored. The driver can also be killed at
+  any time: when it restarts, the finalizer indicates that allocation may be in
+  progress and has to be completed or aborted.
+
+- In a cluster with multiple scheduler instances, two pods might get
+  scheduled concurrently by different schedulers. When they reference
+  the same ResourceClaim which may only get used by one pod at a time,
+  only one pod can be scheduled.
+
+  Both schedulers try to add their pod to the `ReservedFor` field, but only the
+  update that reaches the API server first gets stored. The other one fails
+  with a conflict error and the scheduler which issued it knows that it must
+  put the pod back into the queue, waiting for the ResourceClaim to become
+  usable again.
+
+### Custom parameters
+
+To support arbitrarily complex parameters, both ResourceClass and ResourceClaim
+contain one field of type
+[v1.TypedLocalObjectReference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#typedlocalobjectreference-v1-core)
```

Comment:
Got it. So using it in [IngressBackend [networking/v1]](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#ingressbackend-v1-networking-k8s-io) and [PersistentVolumeClaimSpec [core/v1]](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#persistentvolumeclaimspec-v1-core) should have been avoided, too, I guess. This could still be fixed ("only" breaks Go code, not the JSON encoding), but probably isn't worth it.

Reply:
I was undecided whether I should have a single `ResourceParameterReference` or `ResourceClassParameterReference` and `ResourceClaimParameterReference`. I went for two types because they might become different at some point (my understanding of the reason for the "define your own type" guidance).